doc/sphinx/src/using-closures.rst

```diff
@@ -167,6 +398,23 @@ objects for each equation of state, similar to the vector API for a
 given EOS. Here the indexers/vectors are not over cells, but
 materials.
 
+.. warning::
+
+  It bears repeating: **both the volume fractions and densities act as inputs
+  and outputs**. They are used to define the internal :math:`\overline
+  {\rho}_i` variables at the beginning of the PTE solve. The volume fractions
+  and densities at the end of the PTE solve will represent those for the new
+  PTE state.
+
+.. warning::
+
+  The PTE solvers **require** that all input densities and volume fractions
+  are non-zero. As a result, ``nmat`` refers to the number of *participating*
+  materials. The user is encouraged to wrap their data arrays using an
+  ``Indexer`` concept where, for example, three paricipating PTE materials
+  might be indexed as 5, 7, 20 in the material arrays. This requires overloading
+  the square bracket operator to map from PTE idex to material index.
```

Comment:
My question is, how do we know what the participating materials are? It is easy to say that it is when \overline{rho} is non-zero, but how do we know that \overline{rho} became non-zero in the last time step? Or that \overline{rho} became zero and a material/phase vanished? Is it \overline{rho} that is updated in the timestep? Please educate me.

Reply:
Deciding which materials participate is a very important question, and one we struggle with in an Eulerian frame quite a bit. Many Eulerian codes have some sort of non-zero mass fraction limit that limits participation in the PTE solver, but with a good volume fraction guess it might be better to use a small volume fraction cutoff. In general this is left up to the host codes, but we could imagine tools in the future to aid in this decision. I've seen cases where mixed cells will drive small volume fractions of materials up to densities well over 100 g/cc only for the density to be completely reasonable in the next cycle.

To your second question, the hydro code I think is free to update as it sees fit, i.e. updating something like `$\overline{\rho}$` (which is basically equivalent to updating the masses directly), or maybe the physical densities themselves. I think adopting your suggestion to use `$\overline{\rho}$` as an input would be natural for many hydro codes.

Reply:
I am not sure we want to be in charge of which materials participate and which don't. IMO this is a question for the host code, as different codes want to do different things. In riot, for example, we experimented with this quite a bit. We now iteratively remove the material with the smallest estimated volume fraction until PTE converges.

Reply:
I'm not advocating for a fixed prescription, but just as we provide modifiers to EOS, I think it's reasonable to maybe in the future provide a PTE filter if you will. Something like this could potentially automate the creation of indexer objects for instance.

I'll also note that at some level we already do make this decision in `get_sg_eos()`. We just use zero mass fraction as the cutoff, which is pretty obvious and sane. However, since our PTE solver doesn't support zero mass volume fractions or densities, it may be prudent to generalize this machinery.